xla/service/gpu/runtime/gpublas_lt_matmul.cc

```diff
@@ -62,6 +62,27 @@ XLA_RUNTIME_REGISTER_ENUM_ATTR_DECODING(se::gpu::BlasLt::Epilogue);
 
 namespace gpu {
 
+MatmulPlanVec::Result MatmulPlanVec::GetOrCreate(int device_id, 
+      CreateFunc create_plan) {
+
+  se::gpu::BlasLt::MatmulPlan *plan_raw = nullptr;
+  {
+    absl::MutexLock lock(mu_.get());
+    if(device_id >= plans_.size()) {
+      plans_.resize(device_id + 1);
+    } else {
+      plan_raw = plans_[device_id].get();
+    }
+  }
+  if(plan_raw == nullptr) {
+    VLOG(2) << "Creating new MatmulPlan for device ID: " << device_id;
+    TF_ASSIGN_OR_RETURN(auto plan, create_plan());
+    absl::MutexLock lock(mu_.get());
+    plan_raw = plan.get();
+    plans_[device_id] = std::move(plan);
+  }
```

Comment:
I'm thinking about a possible race condition here.
Is it possible that multiple threads would ask to create a matmul plan for the same device id concurrently?
Because if that's possible, then it would be possible that 2 threads see that plan_raw is nullptr (for the same device_id) and they would both create a new matmul plan and store it.  The problem is that when the 2nd one stores its matmul plan, that would destruct the matmul plan of the first one. So the first one could return a pointer to the destructed matmul plan.

One way to avoid this is to do everything in this function under the protection a single MutexLock?
Would that be ok to do that?

Reply:
huh normally this should not happen because each thread having its own StreamExecutor will be assigned a different **device_id**, but in general, you are right: this solution is not clean when no assumptions about threads is made.. Let me think about it, I did not want to put 'create_plan' under mutex since otherwise all devices executing this matmul op concurrently will have to wait. And who knows how much time cu/hipblas-lt MatmulPlan creation takes..